lib/Solver/SolverCmdLine.cpp

```diff
@@ -112,17 +112,19 @@ cl::opt<bool> UseAssignmentValidatingSolver(
     cl::desc("Debug the correctness of generated assignments (default=false)"),
     cl::cat(SolvingCat));
 
-
-void KCommandLine::HideOptions(llvm::cl::OptionCategory &Category) {
+void KCommandLine::KeepOnlyCategories(
+    std::set<llvm::cl::OptionCategory *> categories) {
   StringMap<cl::Option *> &map = cl::getRegisteredOptions();
 
   for (auto &elem : map) {
     for (auto &cat : elem.second->Categories) {
-      if (cat == &Category) {
+      if (categories.find(cat) == categories.end()) {
         elem.second->setHiddenFlag(cl::Hidden);
       }
     }
   }
```

Comment:
This looks like it will set the element to hidden if it does not have _all_ of the provided categories. I would suggest that we want it to fail if it does not have _any_ of the provided categories (the current behavior probably works in practice, as KLEE options only have KLEE categories):
```
for (auto &elem : map) {
  bool keep = false;
  for (auto &cat : elem.second->Categories) {
    if (categories.find(cat) != categories.end()) {
      keep = true;
      break;
    }
  }
  if (!keep) {
    elem.second->setHiddenFlag(cl::Hidden);
  }
}
```
or, alternatively:
```
for (auto &elem : map) {
  if (!std::any_of(elem.second->Categories.begin(),
                  elem.second->Categories.end(),
                  [](auto const *cat) {
                    return categories.find(cat) != categories.end();
                  })) {
    elem.second->setHiddenFlag(cl::Hidden);
  }
}
```

Reply:
Hi @danielschemmel, the code does this already :) 
It is looking in `categories` (all KLEE categories) and if it doesn't find it there, it sets the flag to hidden.

Reply:
Yes, however, it does so for every category of the option (`for (auto &cat : elem.second->Categories)`).

For example, if we have an option that has two categories: `llvm::cl::GeneralCategory` and `klee::MiscCat`, it would set the hidden flag when encountering `llvm::cl::GeneralCategory`. I feel like this is the wrong behavior and we should keep anything that has any of the given categories.

Reply:
I now see what you mean, but an option can be assigned to a single category: https://llvm.org/docs/CommandLine.html

Reply:
I have tried this (llvm 14), and it is definitely possible to assign multiple categories to an option. Doing so will cause the option to appear multiple times in the help message (once in each category).

We may very well decide that this is not something we wish to support in KLEE, in which case we should just drop the loop altogether and just check against the first element.

Reply:
```
for (auto &elem : map) {
  if (elem.second->Categories.size() != 1 || categories.find(elem.second->Categories[0]) == categories.end()) {
    elem.second->setHiddenFlag(cl::Hidden);
  }
}
```

Reply:
Thanks.  If multiple categories are indeed allowed, let's handle that, as we might want to use this feature.